File: src/core/prekeyStore.js
// X3DH prekey store.
//
// Holds the local device's signed prekey (SPK) and one-time prekeys (OPKs).
// The bundle we hand out to remote peers is:
//   { IK_pub, SPK_id, SPK_pub, SPK_sig, OPK: { id, pub } }
// where SPK_sig = ECDSA(identityKey, SPK_pub_spki).
//
// Private key material is kept non-extractable inside IndexedDB; only the
// SPKI-serialized public halves ever leave the device. This mirrors how
// `identityKey.js` stores the long-term ECDSA key.
//
// SPK lifecycle:
//   - exactly one SPK is `active` at a time (the one we publish);
//   - when rotated, the previous SPK is kept for a grace window so late
//     inbound messages encrypted to it can still decrypt;
//   - `pruneRetiredSPKs(maxAgeMs)` removes ones older than the window.
//
// OPK lifecycle:
//   - pool of N unused OPKs;
//   - `consumeOPK(id)` flips `used=true` so the same OPK is never reused
//     (X3DH forward secrecy depends on this);
//   - `pruneUsedOPKs(maxAgeMs)` removes spent entries after a grace window.

import { openDatabase } from './db.js';
import { signBytes } from './identityKey.js';

const SPK_ROTATION_MS = 7 * 24 * 60 * 60 * 1000; // 7 days
const USED_OPK_RETENTION_MS = 24 * 60 * 60 * 1000; // 1 day
const DEFAULT_OPK_POOL_SIZE = 100;

function randomId(prefix) {
  const bytes = new Uint8Array(8);
  crypto.getRandomValues(bytes);
  let hex = '';
  for (const b of bytes) hex += b.toString(16).padStart(2, '0');
  return `${prefix}-${hex}`;
}

async function generateECDHPair() {
  return crypto.subtle.generateKey(
    { name: 'ECDH', namedCurve: 'P-256' },
    false,
    ['deriveBits']
  );
}

async function exportSpki(publicKey) {
  const buf = await crypto.subtle.exportKey('spki', publicKey);
  return new Uint8Array(buf);
}

/**
 * Generate a fresh signed prekey, sign its SPKI with the local identity, and
 * persist it as the new active SPK. The previous active SPK, if any, is
 * demoted to `retired` so late inbound messages can still decrypt.
 */
export async function rotateSignedPrekey() {
  const db = await openDatabase();
  const pair = await generateECDHPair();
  const pubSpki = await exportSpki(pair.publicKey);
  const sig = await signBytes(pubSpki);

  const id = randomId('spk');
  const record = {
    id,
    kind: 'spk',
    status: 'active',
    used: 0,
    privateKey: pair.privateKey,
    pubSpki,
    sig,
    createdAt: Date.now()
  };

  const tx = db.transaction('prekeys', 'readwrite');
  const store = tx.objectStore('prekeys');
  // Demote any currently-active SPK.
  const idx = store.index('kind');
  let cursor = await idx.openCursor('spk');
  while (cursor) {
    const v = cursor.value;
    if (v.status === 'active') {
      await cursor.update({ ...v, status: 'retired', retiredAt: Date.now() });
    }
    cursor = await cursor.continue();
  }
  await store.put(record);
  await tx.done;
  return { id, pubSpki, sig };
}

/** Returns the active SPK record or null. */
export async function getActiveSignedPrekey() {
  const db = await openDatabase();
  const tx = db.transaction('prekeys', 'readonly');
  const idx = tx.objectStore('prekeys').index('kind');
  let cursor = await idx.openCursor('spk');
  while (cursor) {
    if (cursor.value.status === 'active') return cursor.value;
    cursor = await cursor.continue();
  }
  return null;
}

/** Look up any SPK by id — active or retired. Used by recipients on inbound. */
export async function getSignedPrekeyById(id) {
  const db = await openDatabase();
  const rec = await db.get('prekeys', id);
  return rec && rec.kind === 'spk' ? rec : null;
}

/** Drop retired SPKs older than `maxAgeMs` (default 14 days). */
export async function pruneRetiredSPKs(maxAgeMs = 14 * 24 * 60 * 60 * 1000) {
  const db = await openDatabase();
  const cutoff = Date.now() - maxAgeMs;
  const tx = db.transaction('prekeys', 'readwrite');
  const idx = tx.objectStore('prekeys').index('kind');
  let cursor = await idx.openCursor('spk');
  let removed = 0;
  while (cursor) {
    const v = cursor.value;
    if (v.status === 'retired' && Number(v.retiredAt || 0) < cutoff) {
      await cursor.delete();
      removed += 1;
    }
    cursor = await cursor.continue();
  }
  await tx.done;
  return removed;
}

/**
 * Generate `count` fresh OPKs and persist them. Returns the freshly-minted
 * `{ id, pubSpki }[]` so callers can include them in an outgoing bundle.
 */
export async function generateOneTimePrekeys(count = DEFAULT_OPK_POOL_SIZE) {
  // Mint all key material up front. IDB transactions close across
  // non-IDB awaits, so we can't interleave `generateKey` with `store.put`.
  const minted = [];
  for (let i = 0; i < count; i++) {
    const pair = await generateECDHPair();
    const pubSpki = await exportSpki(pair.publicKey);
    minted.push({
      id: randomId('opk'),
      kind: 'opk',
      status: 'fresh',
      used: 0,
      privateKey: pair.privateKey,
      pubSpki,
      createdAt: Date.now()
    });
  }

  const db = await openDatabase();
  const tx = db.transaction('prekeys', 'readwrite');
  const store = tx.objectStore('prekeys');
  const writes = minted.map((rec) => store.put(rec));
  await Promise.all(writes);
  await tx.done;
  return minted.map(({ id, pubSpki }) => ({ id, pubSpki }));
}

/** Count of unused OPKs remaining in the pool. */
export async function countFreshOPKs() {
  const db = await openDatabase();
  const tx = db.transaction('prekeys', 'readonly');
  const idx = tx.objectStore('prekeys').index('kind_used');
  return await idx.count(['opk', 0]);
}

/**
 * Pull public parts of up to `n` fresh OPKs for publication in a bundle.
 * Does not mark them used — consumption happens on inbound X3DH via
 * `consumeOPK(id)`.
 */
export async function listFreshOPKs(n = 20) {
  const db = await openDatabase();
  const tx = db.transaction('prekeys', 'readonly');
  const idx = tx.objectStore('prekeys').index('kind_used');
  let cursor = await idx.openCursor(IDBKeyRange.only(['opk', 0]));
  const out = [];
  while (cursor && out.length < n) {
    out.push({ id: cursor.value.id, pubSpki: cursor.value.pubSpki });
    cursor = await cursor.continue();
  }
  return out;
}

/**
 * Retrieve a fresh OPK record and mark it used atomically. Returns the record
 * (including privateKey) so the caller can immediately perform DH. If the OPK
 * is missing or already used, returns null.
 */
export async function consumeOPK(id) {
  const db = await openDatabase();
  const tx = db.transaction('prekeys', 'readwrite');
  const store = tx.objectStore('prekeys');
  const rec = await store.get(id);
  if (!rec || rec.kind !== 'opk' || rec.used !== 0) {
    await tx.done;
    return null;
  }
  await store.put({ ...rec, used: 1, usedAt: Date.now() });
  await tx.done;
  return rec;
}

/** Remove OPKs marked used longer than `maxAgeMs` ago. */
export async function pruneUsedOPKs(maxAgeMs = USED_OPK_RETENTION_MS) {
  const db = await openDatabase();
  const cutoff = Date.now() - maxAgeMs;
  const tx = db.transaction('prekeys', 'readwrite');
  const idx = tx.objectStore('prekeys').index('kind_used');
  let cursor = await idx.openCursor(IDBKeyRange.only(['opk', 1]));
  let removed = 0;
  while (cursor) {
    const v = cursor.value;
    if (Number(v.usedAt || 0) < cutoff) {
      await cursor.delete();
      removed += 1;
    }
    cursor = await cursor.continue();
  }
  await tx.done;
  return removed;
}

/**
 * Top-level bootstrap: ensure there is an active SPK newer than the rotation
 * window, and top the OPK pool up to `targetPool`. Call once at startup and
 * whenever the user publishes/refreshes their bundle.
 */
export async function ensurePrekeysReady({
  rotationMs = SPK_ROTATION_MS,
  targetPool = DEFAULT_OPK_POOL_SIZE,
  minPool = 20
} = {}) {
  const active = await getActiveSignedPrekey();
  const needsRotation = !active || Date.now() - Number(active.createdAt || 0) > rotationMs;
  if (needsRotation) {
    await rotateSignedPrekey();
  }
  const count = await countFreshOPKs();
  if (count < minPool) {
    await generateOneTimePrekeys(targetPool - count);
  }
}

File: lib/core/prekey_store.dart
// Port of src/core/prekeyStore.js — SPK + OPK management for X3DH.
//
// Bundle we hand out:
//   { IK_pub, SPK{id, pub, sig}, OPK{id, pub}? }
// where SPK_sig = ECDSA(identityKey) over SPK_pub_spki.
//
// SPK lifecycle:
//   - exactly one `active` SPK at a time (the one we publish);
//   - on rotation the previous active is demoted to `retired` for a grace
//     window so late inbound still decrypts;
//   - [pruneRetiredSPKs] drops ones older than the window.
//
// OPK lifecycle:
//   - pool of N unused OPKs;
//   - [consumeOPK] flips `used=true` — X3DH forward secrecy depends on
//     never reusing an OPK;
//   - [pruneUsedOPKs] drops spent entries after a short grace window.

import 'dart:math';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

import 'identity_key.dart' as identity_key;
import 'key_store.dart';
import 'spki_codec.dart';

const String _prekeysTable = 'prekeys';

const int _spkRotationMs = 7 * 24 * 60 * 60 * 1000; // 7 days
const int _usedOpkRetentionMs = 24 * 60 * 60 * 1000; // 1 day
const int _defaultOpkPoolSize = 100;
const int _retiredSpkGraceMs = 14 * 24 * 60 * 60 * 1000;

final _ecdh = Ecdh.p256(length: 32);
final _secureRandom = Random.secure();

String _randomId(String prefix) {
  final bytes = List<int>.generate(8, (_) => _secureRandom.nextInt(256));
  final hex = bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  return '$prefix-$hex';
}

Future<EcKeyPair> _generateEcdhPair() => _ecdh.newKeyPair();

Future<Uint8List> _exportSpki(EcKeyPair pair) async {
  final pub = await pair.extractPublicKey();
  if (pub is! EcPublicKey) {
    throw StateError('prekey: expected EcPublicKey');
  }
  return buildP256Spki(x: pub.x, y: pub.y);
}

Future<Uint8List> _extractPriv(EcKeyPair pair) async =>
    // cryptography 2.9: `extractPrivateKeyBytes` is SimpleKeyPair-only.
    // For EcKeyPair, extract to data then read the scalar `d`.
    Uint8List.fromList((await pair.extract()).d);

/// Public info about a signed prekey.
class SignedPrekey {
  const SignedPrekey({
    required this.id,
    required this.pubSpki,
    required this.sig,
  });
  final String id;
  final Uint8List pubSpki;
  final Uint8List sig;
}

/// Full SPK record — what gets persisted and what X3DH responders load to DH
/// against incoming hellos.
class SignedPrekeyRecord {
  const SignedPrekeyRecord({
    required this.id,
    required this.status,
    required this.pubSpki,
    required this.sig,
    required this.privateKey,
    required this.createdAt,
    this.retiredAt,
  });
  final String id;
  final String status; // 'active' | 'retired'
  final Uint8List pubSpki;
  final Uint8List sig;
  final EcKeyPair privateKey;
  final int createdAt;
  final int? retiredAt;
}

Future<SignedPrekeyRecord?> _rowToSpk(Map<String, Object?> row) async {
  if (row['kind'] != 'spk') return null;
  final privBytes = row['privBytes'];
  final pubSpki = row['pubSpki'];
  final sig = row['sig'];
  if (privBytes is! List<int> || pubSpki is! List<int> || sig is! List<int>) {
    return null;
  }
  final point = parseP256Spki(pubSpki);
  // EcKeyPairData constructor wants d/x/y split (see identity_key.dart for
  // the migration notes from cryptography 2.7 → 2.9).
  final keyPair = EcKeyPairData(
    d: privBytes,
    x: point.x,
    y: point.y,
    type: KeyPairType.p256,
  );
  return SignedPrekeyRecord(
    id: row['id'] as String,
    status: (row['status'] as String?) ?? 'retired',
    pubSpki: Uint8List.fromList(pubSpki),
    sig: Uint8List.fromList(sig),
    privateKey: keyPair,
    createdAt: (row['createdAt'] as num?)?.toInt() ?? 0,
    retiredAt: (row['retiredAt'] as num?)?.toInt(),
  );
}

/// Generate a fresh SPK, sign it with the local identity, and persist as
/// the new active SPK. Any prior active SPK gets demoted to `retired`.
Future<SignedPrekey> rotateSignedPrekey() async {
  final pair = await _generateEcdhPair();
  final pubSpki = await _exportSpki(pair);
  final sig = await identity_key.signBytes(pubSpki);
  final privBytes = await _extractPriv(pair);
  final id = _randomId('spk');

  final all = await keyStore()
      .getAll(_prekeysTable, indexField: 'kind', indexValue: 'spk');
  for (final r in all) {
    if (r['status'] == 'active') {
      await keyStore().put(_prekeysTable, {
        ...r,
        'status': 'retired',
        'retiredAt': DateTime.now().millisecondsSinceEpoch,
      });
    }
  }

  await keyStore().put(_prekeysTable, {
    'id': id,
    'kind': 'spk',
    'status': 'active',
    'used': 0,
    'privBytes': privBytes,
    'pubSpki': pubSpki,
    'sig': sig,
    'createdAt': DateTime.now().millisecondsSinceEpoch,
  });

  return SignedPrekey(id: id, pubSpki: pubSpki, sig: sig);
}

Future<SignedPrekeyRecord?> getActiveSignedPrekey() async {
  final rows = await keyStore()
      .getAll(_prekeysTable, indexField: 'kind', indexValue: 'spk');
  for (final row in rows) {
    if (row['status'] == 'active') return _rowToSpk(row);
  }
  return null;
}

Future<SignedPrekeyRecord?> getSignedPrekeyById(String id) async {
  final row = await keyStore().get(_prekeysTable, id);
  if (row == null) return null;
  return _rowToSpk(row);
}

Future<int> pruneRetiredSPKs({int maxAgeMs = _retiredSpkGraceMs}) async {
  final cutoff = DateTime.now().millisecondsSinceEpoch - maxAgeMs;
  final rows = await keyStore()
      .getAll(_prekeysTable, indexField: 'kind', indexValue: 'spk');
  var removed = 0;
  for (final row in rows) {
    if (row['status'] == 'retired') {
      final retiredAt = (row['retiredAt'] as num?)?.toInt() ?? 0;
      if (retiredAt < cutoff) {
        await keyStore().delete(_prekeysTable, row['id'] as String);
        removed++;
      }
    }
  }
  return removed;
}

// ─── OPKs ──────────────────────────────────────────────────────────

/// Public one-time prekey (id + SPKI).
class OneTimePrekey {
  const OneTimePrekey({required this.id, required this.pubSpki});
  final String id;
  final Uint8List pubSpki;
}

/// Full OPK record — what the responder loads to perform DH4.
class OneTimePrekeyRecord {
  const OneTimePrekeyRecord({
    required this.id,
    required this.pubSpki,
    required this.privateKey,
    required this.used,
    required this.createdAt,
    this.usedAt,
  });
  final String id;
  final Uint8List pubSpki;
  final EcKeyPair privateKey;
  final int used;
  final int createdAt;
  final int? usedAt;
}

Future<OneTimePrekeyRecord?> _rowToOpk(Map<String, Object?> row) async {
  if (row['kind'] != 'opk') return null;
  final privBytes = row['privBytes'];
  final pubSpki = row['pubSpki'];
  if (privBytes is! List<int> || pubSpki is! List<int>) return null;
  final point = parseP256Spki(pubSpki);
  final keyPair = EcKeyPairData(
    d: privBytes,
    x: point.x,
    y: point.y,
    type: KeyPairType.p256,
  );
  return OneTimePrekeyRecord(
    id: row['id'] as String,
    pubSpki: Uint8List.fromList(pubSpki),
    privateKey: keyPair,
    used: (row['used'] as num?)?.toInt() ?? 0,
    createdAt: (row['createdAt'] as num?)?.toInt() ?? 0,
    usedAt: (row['usedAt'] as num?)?.toInt(),
  );
}

/// Mint [count] fresh OPKs and persist them. Returns their public id/SPKI
/// pairs so the caller can drop them into an outgoing bundle.
Future<List<OneTimePrekey>> generateOneTimePrekeys(
    [int count = _defaultOpkPoolSize]) async {
  final minted = <OneTimePrekey>[];
  final now = DateTime.now().millisecondsSinceEpoch;
  for (var i = 0; i < count; i++) {
    final pair = await _generateEcdhPair();
    final pubSpki = await _exportSpki(pair);
    final privBytes = await _extractPriv(pair);
    final id = _randomId('opk');
    await keyStore().put(_prekeysTable, {
      'id': id,
      'kind': 'opk',
      'status': 'fresh',
      'used': 0,
      'privBytes': privBytes,
      'pubSpki': pubSpki,
      'createdAt': now,
    });
    minted.add(OneTimePrekey(id: id, pubSpki: pubSpki));
  }
  return minted;
}

Future<int> countFreshOPKs() async {
  final rows = await keyStore()
      .getAll(_prekeysTable, indexField: 'kind', indexValue: 'opk');
  var n = 0;
  for (final r in rows) {
    if ((r['used'] as num?)?.toInt() == 0) n++;
  }
  return n;
}

/// Pull up to [n] fresh OPKs for publication. Does **not** mark used —
/// consumption happens in [consumeOPK] on inbound X3DH.
Future<List<OneTimePrekey>> listFreshOPKs([int n = 20]) async {
  final rows = await keyStore()
      .getAll(_prekeysTable, indexField: 'kind', indexValue: 'opk');
  final out = <OneTimePrekey>[];
  for (final r in rows) {
    if (out.length >= n) break;
    if ((r['used'] as num?)?.toInt() == 0) {
      out.add(OneTimePrekey(
        id: r['id'] as String,
        pubSpki: Uint8List.fromList(r['pubSpki'] as List<int>),
      ));
    }
  }
  return out;
}

/// Fetch an OPK and mark it used in one step. Returns null if missing or
/// already consumed — a second consume of the same id is a no-op so replays
/// cannot force key reuse.
Future<OneTimePrekeyRecord?> consumeOPK(String id) async {
  final row = await keyStore().get(_prekeysTable, id);
  if (row == null || row['kind'] != 'opk') return null;
  if ((row['used'] as num?)?.toInt() != 0) return null;
  final rec = await _rowToOpk(row);
  if (rec == null) return null;
  await keyStore().put(_prekeysTable, {
    ...row,
    'used': 1,
    'usedAt': DateTime.now().millisecondsSinceEpoch,
  });
  return rec;
}

Future<int> pruneUsedOPKs({int maxAgeMs = _usedOpkRetentionMs}) async {
  final cutoff = DateTime.now().millisecondsSinceEpoch - maxAgeMs;
  final rows = await keyStore()
      .getAll(_prekeysTable, indexField: 'kind', indexValue: 'opk');
  var removed = 0;
  for (final row in rows) {
    if ((row['used'] as num?)?.toInt() == 1) {
      final usedAt = (row['usedAt'] as num?)?.toInt() ?? 0;
      if (usedAt < cutoff) {
        await keyStore().delete(_prekeysTable, row['id'] as String);
        removed++;
      }
    }
  }
  return removed;
}

/// One-shot bootstrap: ensure the SPK isn't stale and top up the OPK pool.
/// Call at startup and whenever the bundle is refreshed.
Future<void> ensurePrekeysReady({
  int rotationMs = _spkRotationMs,
  int targetPool = _defaultOpkPoolSize,
  int minPool = 20,
}) async {
  final active = await getActiveSignedPrekey();
  final now = DateTime.now().millisecondsSinceEpoch;
  final needsRotation = active == null || (now - active.createdAt) > rotationMs;
  if (needsRotation) {
    await rotateSignedPrekey();
  }
  final count = await countFreshOPKs();
  if (count < minPool) {
    await generateOneTimePrekeys(targetPool - count);
  }
}
